Updated Usage with working code

RichardLitt opened the pull request:
I was not able to get it working as it was; this should help a bit. See https://github.com/ipfs/interface-ipfs-core/issues/43. Also, updated ToC.

dignifiedquire commented on README.md:
```diff
@@ -75,22 +60,42 @@ In Go land:
+'use strict'
+
+const test = require('interface-ipfs-core')
+const IPFS = require('ipfs')
```
it should be made clear that this does should be your own ipfs instance and not the package `ipfs`.

dignifiedquire commented on README.md:
```diff
@@ -75,22 +60,42 @@ In Go land:
 
-// use all of the test suits
-test.all(common)
+test.files(common)
```
This is only an example of running the `files` tests, not all of them, this should be noted with a reference to where one can find a list of all of them

RichardLitt replied:
Where would that be?

dignifiedquire replied:
here: https://github.com/ipfs/interface-ipfs-core/blob/master/src/index.js

dignifiedquire commented on package.json:
```diff
@@ -36,7 +36,8 @@
   "devDependencies": {
-    "aegir": "^3.2.0"
+    "aegir": "^3.2.0",
+    "ipfs": "^0.14.1"
```
This is not a good idea, this adds the dependency to all the repos requiring it even though they might not need it

dignifiedquire commented on test.js:
```diff
@@ -0,0 +1,19 @@
+  }
+}
+
+test.files(common)
```
This should be in something like an `examples` folder or even in another repo, as this is only for show.

RichardLitt replied:
Moved.

dignifiedquire commented on README.md:
```diff
@@ -17,22 +18,6 @@
-      - [`object.patch.addLink`](/API/object#objectpatchaddlink)
-      - [`object.patch.rmLink`](/API/object#objectpatchrmlink)
-      - [`object.patch.appendData`](/API/object#objectpatchappenddata)
-      - [`object.patch.setData`](/API/object#objectpatchsetdata)
```
Why did you remove all of these?

RichardLitt replied:
They are not in the file. The Table of Contents should reflect the current file.

diasdavid replied:
It is the table of contents of the repo, which includes the several APIs spec definitions. Otherwise how will devs be able to open the repo and jump quickly to where they need?

dignifiedquire commented on README.md:
````diff
@@ -75,22 +60,42 @@ In Go land:
-```
-var test = require('interface-ipfs-core')
+```js
+/* eslint-env mocha */
````
This is not needed, as neither `describe` nor `it` are used in this code snippet.

dignifiedquire commented on README.md:
```diff
@@ -75,22 +60,42 @@ In Go land:
+'use strict'
+
+const test = require('interface-ipfs-core')
+const IPFS = require('ipfs') // This should be your own instance, not the package ipfs.
```
Maybe sth like `const IPFS = getYourIPFSInstance()` instead?

diasdavid replied:
👍

RichardLitt replied:
Can we make this process cleaner? How do I set up my own instance, easily?

dignifiedquire commented on README.md:
```diff
@@ -75,22 +60,42 @@ In Go land:
   setup: function (cb) {
-    cb(null, yourIPFSInstance)
+    const ipfs = new IPFS()
+    ipfs.load(() => {
```
`ipfs.load` is specific to `js-ipfs` so I am not sure it makes sense to have in here.

RichardLitt replied:
Should I just remove it?

diasdavid replied:
yes please :)

dignifiedquire commented on README.md:
````diff
@@ -75,22 +60,42 @@ In Go land:
+Then, using `mocha` (for Node.js) or a test runner with compatible API, do:
+
+```sh
+mocha test.js
````
Caan you add `$` in front please, as well as below

diasdavid replied:
or a `>`

dignifiedquire commented on examples/test.js:
```diff
@@ -0,0 +1,19 @@
+/* eslint-env mocha */
```
Why the duplication of this file, with the one in readme?

RichardLitt replied:
I want to be able to run it, immediately. You can't do that if it is stuck in a README.

diasdavid replied:
There isn't however (or will be) an implementation of IPFS inside this repo.

RichardLitt replied:
There still ought to be a way to run the code, shouldn't there? Otherwise it's not a usable example. See title of the PR.

diasdavid commented:
@RichardLitt are you familiar with:

- https://github.com/maxogden/abstract-blob-store
- https://github.com/libp2p/interface-transport
- https://github.com/libp2p/interface-connection
- https://github.com/libp2p/interface-stream-muxer
- https://github.com/libp2p/interface-peer-routing
- https://github.com/libp2p/interface-record-store
- https://github.com/libp2p/interface-connection
- https://github.com/libp2p/interface-libp2p

These are all spec definitions plus tests that can be used to make sure that the implementations are compliant.

It would actually be good if https://github.com/ipfs/http-api-spec would have a battery of tests that could be used to test the API, so that we could run them against go-ipfs and js-ipfs. @whyrusleeping how interesting would it be to you to have the CLI extracted from go-ipfs, so that it can be compiled as a standalone binary so that we can put the shardness tests available to run against any IPFS implementation? (Might be better to talk about this in the http-api-spec repo)

RichardLitt commented:
So it's common to not have runnable usage code? Can we compromise by adding a paragraph explaining out how to set up your own IPFS Instance that can be called programmatically? How would I do that?

RichardLitt commented:
And yes, please raise an issue about that in http-api-spec. I've been wanting that for a long time, but I can't do it alone.

diasdavid commented:
https://github.com/ipfs/interface-ipfs-core/pull/44#issuecomment-237845345

Yes, it is common not to have runnable code. Note that these tests are also ran in js-ipfs-api, which doesn't spawn the node at all.

https://github.com/ipfs/interface-ipfs-core/pull/44#issuecomment-237845417

There you go: https://github.com/ipfs/http-api-spec/issues/113

RichardLitt commented:
Thanks. Ok, changed to yourIPFSInstance, removed example code. 

I still think we need a paragraph on how to spawn your own instance. Can you describe that to me, in simple writing, here?

RichardLitt commented:
@diasdavid I'd like to figure how how to spawn an instance, so I can write a better doc here about how to do that. Where should I ask that? On IRC?

diasdavid commented on README.md:
```diff
@@ -76,22 +76,38 @@ In Go land:
+'use strict'
+
+const test = require('interface-ipfs-core')
+const yourIPFSInstance = require('ipfs')
```
This doesn't yield an IPFS instance.

diasdavid commented:
> @diasdavid I'd like to figure how to spawn an instance.

That is the thing, it is not up to the 'spec' to define how to spawn an IPFS instance. The tests only tell you that you have a 'correct' IPFS instance if the thing you pass to it is an IPFS instance that has all the functions and behaviour the tests expect.

RichardLitt commented:
But you can't run the tests if you don't know how to run an instance. Isn't it important to tell people how to run the tests?

diasdavid commented:
> But you can't run the tests if you don't know how to run an instance. Isn't it important to tell people how to run the tests?

Not quite, if you want to use this test suite to test 'your own IPFS interface implementation' and you don't know how to create an instance from your implementation, you are in a deeper problem.

Note that this tests get executed in js-ipfs and js-ipfs-api repos, as they are required (which is why the tests are in the src folder, because this is a test suite module). 

Hope this clarifies, closing this PR.
